drv/psc-seq-server/src/main.rs

```diff
@@ -3,48 +3,586 @@
+                ok,
+            ) => {
+                if turn_on_deadline <= now {
+                    match ok {
+                        Status::NotGood => {
+                            // Do it all again then, I guess.
+                            let turn_on_deadline =
+                                now.wrapping_add(FAULT_OFF_MS);
+                            self.state =
+                                PsuState::Present(PresentState::Faulted {
+                                    turn_on_deadline,
+                                });
+                            Some(ActionRequired::NoteStickyFault)
+                        }
+                        Status::Good => {
+                            self.state = PsuState::Present(PresentState::On);
+                            Some(ActionRequired::EnableMe)
+                        }
+                    }
+                } else {
+                    // Remain in this state.
+                    None
+                }
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone)]
+#[must_use]
+enum ActionRequired {
+    /// Requests that this PSU be enabled by setting the corresponding
+    /// `ENABLE_L` low.
+    EnableMe,
+    /// Requests that this PSU be disabled by setting the corresponding
+    /// `ENABLE_L` high. `attempt_snapshot` will be `true` if the PSU is
+    /// believed to still be present and recording data may be useful, or
+    /// `false` if the PSU is believed removed and isn't worth polling.
+    DisableMe { attempt_snapshot: bool },
+    /// Emitted when a PSU is disabled due to a fault, has gone through the
+    /// fault recovery process, but is still indicating a fault. The caller
+    /// may wish to log this condition, or may not, since it will appear
+    /// every `FAULT_MS` milliseconds and could dominate the logs.
+    NoteStickyFault,
+}
+
+#[derive(Copy, Clone)]
+enum PsuState {
+    /// The PSU is detected as not present. In this state, we cannot trust the
+    /// OK signal, and we deassert the ENABLE signal.
+    NotPresent,
+    /// The PSU is detected as present. We assume this at powerup until proven
+    /// otherwise.
+    Present(PresentState),
+}
+
+#[derive(Copy, Clone)]
+enum PresentState {
+    /// We are allowing the ON signal to float active (low).
+    ///
+    /// This is the initial state upon either detecting a new PSU, or power
+    /// up/restart in cases where the PSU is not forced off.
+    ///
+    /// We will exit this state if the OK line is pulled low, or if we detect a
+    /// fault.
+    On,
+
+    /// The PSU has just appeared and we're waiting a bit to confirm that it's
+    /// stable before turning it on. (Waiting in this state provides some
+    /// debouncing for contact scrape.)
+    NewlyInserted { settle_deadline: u64 },
+    /// The PSU has unexpectedly deasserted the OK signal, or failed to assert
+    /// it within a reasonable amount of time after being turned on.
+    Faulted { turn_on_deadline: u64 },
+}
```

Comment:
nit, take it or leave it: I might consider defining these enums _before_ the logic that operates on them, rather than after --- the documentation here is valuable context when reading the implementations of those methods, so it might be nice if a top-to-bottom reader encountered it first? not a big deal though.

Reply:
Done!